### ISSUE TITLE
bug in VRM?

### ISSUE THREAD

**Issue by dschneiderch** [jeffreyevans/spatialEco#2]
I think the cos and sin in L33-34 are flipped from what they should be. These are the xy and z vectors.

currently, I get substantially different results from SAGA GIS but if I flip the definitions for the xy and z vectors then this function and SAGA implementaion are in agreement with each other.

https://github.com/jeffreyevans/spatialEco/blob/8fc07e88d359b14d9e611f713ccbc7fc48300e83/R/vrm.R#L33-L34

**Comment by jeffreyevans** [comment]
Indeed, the cos and sin functions were flipped. Fixed now, thanks for catching this.

**Comment by dschneiderch** [comment]
Thanks for the very handy package!

**Comment by r2j2ritson** [comment]
Hi, I have just been looking through Sappington's 2010 Python script (used for his ArcGIS Add-in, https://www.arcgis.com/home/item.html?id=9e4210b3ee7b413bbb1f98fb9c5b22d4) and noticed his x and y rasters are calculated like so:
`gp.SingleOutputMapAlgebra_sa("con(" + AspectRaster + " == -1, 0, sin(" + AspectRasterRad + ") * " + xyRaster + ")", xRaster)`
`gp.SingleOutputMapAlgebra_sa("con(" + AspectRaster + " == -1, 0, cos(" + AspectRasterRad + ") * " + xyRaster + ")", yRaster)`

Which would result in different values than the currently in SpatialEco::vrm since all values of '-1' are set to '0' instead of taking the sine of '-1' like is currently being done:
>https://github.com/jeffreyevans/spatialEco/blob/8fc07e88d359b14d9e611f713ccbc7fc48300e83/R/vrm.R#:~:text=sin.asp%20%3C%2D%20raster,sin.slp%20%20%20%20%23%20xRaster

My preliminary results have produced rasters different looking than those produced by this function versus Sappington's ArcGIS add-in. Has anyone else using Sappington's VRM noticed major differences?

**Comment by jeffreyevans** [comment]
Robert,
I will take a look at this a bit deeper next week but, I believe you are correct in what you are seeing with the recoding of -1 to 0 before taking the sin and cos. This is because ArcGIS codes flat areas, in aspect, with a -1 whereas, the raster::terrain function does not rather, outputting NA’s. As such, there is no need to convert -1 values in R and flat areas can just be set to a zero constant down the road.

Best,
Jeff

Jeffrey S. Evans, Ph.D., | Senior Landscape Ecologist & Biometrician
The Nature Conservancy | Protected Lands Science
Visiting Professor | University of Wyoming | Zoology & Physiology
Laramie, WY | ***@***.******@***.***> | (970) 672-6766<tel:(970)%20672-6766>

**Comment by r2j2ritson** [comment]
Thanks Jeff! I may have jumped the gun a little. Still trying to figure out why my R outputs for VRM differ from the ArcGIS outputs.

**Comment by jeffreyevans** [comment]
Robert,
Not sure about the difference but, I have to say that I have never cared for this particular metric. I put it in the package by request but, have neve found it to be a stable metric and honestly, there is something amiss with the algebra behind it. That said, I am more than happy to figure out why there is a difference between ArcGIS and R outputs. When I get a change I will work through the Python script you provided and check each interim output against what R produces to see if there is a mismatch somewhere.

Have you checked what type of differences you are getting? Are the distributions the same but in a difference scale? Or, are they totally different?

Best,
Jeff

**Comment by jeffreyevans** [comment]
Oh, and BTW, this syntax;

gp.SingleOutputMapAlgebra_sa("con(" + AspectRaster + " == -1, 0, sin(" + AspectRasterRad + ") * " + xyRaster + ")", xRaster)

I do not believe would yield correct results because cos(0) and sin(0) would be invalid in ArcGIS and return NoData. In R, error scoping returns 0 and 1 respectively but, even this is new behavior and it use to return inf. Back in the day, with workstation AML and early ArcPy, I would just code -1 to a very small number as a way around this issue. Perhaps ESRI is now parsing all its math functions through Python but, I would certainly not count on that.

-jeff

**Comment by r2j2ritson** [comment]
Jeff,
A biologist I work with wants to use it in an RSF. I created a vrm raster using the python code/arcgis tool for one of their study areas but they decided they wanted a statewide raster (idaho). I moved to R because the larger raster kept crashing on me (and R is my stronger coding language). I haven't checked the distributions explicitly yet, but the differences in scale caught my attention to look into this. I fixed some minor bugs in my initial code and retrying. I'll let you know what I come up with.
I'm rudimentary  with Python and trying to make sure I didn't miss anything with the R translation.
